transports/tcp/src/lib.rs

```diff
@@ -194,40 +193,30 @@ impl Transport for TcpConfig {
     }
 
     fn nat_traversal(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
-        // Check that `server` only has two components and retreive them.
-        let mut server_protocols_iter = server.iter();
-        let server_proto1 = server_protocols_iter.next()?;
-        let server_proto2 = server_protocols_iter.next()?;
-        if server_protocols_iter.next().is_some() {
-            return None;
-        }
+        let mut address = Multiaddr::default();
+
+        let mut server_iter = server.iter();
+        let mut observed_iter = observed.iter();
 
-        // Check that `observed` only has two components and retreive them.
-        let mut observed_protocols_iter = observed.iter();
-        let observed_proto1 = observed_protocols_iter.next()?;
-        let observed_proto2 = observed_protocols_iter.next()?;
-        if observed_protocols_iter.next().is_some() {
-            return None;
+        // Use the observed IP address.
+        match (server_iter.next(), observed_iter.next()) {
+            (Some(Protocol::Ip4(_)), Some(x@Protocol::Ip4(_))) => address.append(x),
+            (Some(Protocol::Ip6(_)), Some(x@Protocol::Ip6(_))) => address.append(x),
+            _ => return None
         }
 
-        // Check that `server` is a valid TCP/IP address.
-        match (&server_proto1, &server_proto2) {
-            (&Protocol::Ip4(_), &Protocol::Tcp(_))
-            | (&Protocol::Ip6(_), &Protocol::Tcp(_)) => {}
-            _ => return None,
+        // Check for TCP but retain the server port.
```

Comment:
I think we don't need the observed address to have the TCP suffix. Just matching IP is enough.